tower-http/src/add_extension.rs

```diff
@@ -0,0 +1,51 @@
+use crate::common::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct AddExtensionLayer<T> {
+    value: T,
+}
+
+impl<T> AddExtensionLayer<T> {
+    pub fn new(value: T) -> Self {
+        AddExtensionLayer { value }
+    }
+}
+
+impl<S, T> Layer<S> for AddExtensionLayer<T>
+where
+    T: Clone,
+{
+    type Service = AddExtension<S, T>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        AddExtension {
+            inner,
+            value: self.value.clone(),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct AddExtension<S, T> {
+    inner: S,
+    value: T,
+}
+
+impl<ResBody, S, T> Service<Request<ResBody>> for AddExtension<S, T>
+where
+    S: Service<Request<ResBody>>,
+    T: 'static + Clone + Send + Sync,
```

Comment:
minor nit: but I prefer this style 
```suggestion
    T: Clone + Send + Sync + 'static,
```

The order I like is Non auto trait > auto trait > lifetime